src/node_locks.cc

```diff
@@ -0,0 +1,738 @@
+            held_locks_.erase(held_locks_iter);
+          }
+        }
+      }
+
+      // Wake other environments
+      for (Environment* target_env : envs_to_notify) {
+        if (target_env != env) {
+          WakeEnvironment(target_env);
+        }
+      }
+    }
+
+    // Create and store the new granted lock
+    auto granted_lock =
+        std::make_shared<Lock>(env,
+                               grantable_request->name(),
+                               grantable_request->mode(),
+                               grantable_request->client_id(),
+                               grantable_request->waiting_promise(),
+                               grantable_request->released_promise());
+    {
+      Mutex::ScopedLock scoped_lock(mutex_);
+      held_locks_[grantable_request->name()].push_back(granted_lock);
+    }
+
+    // Call user callback
+    Local<Object> lock_info_obj =
+        CreateLockInfoObject(isolate,
+                             context,
+                             grantable_request->name(),
+                             grantable_request->mode(),
+                             grantable_request->client_id());
+    if (lock_info_obj.IsEmpty()) {
+      return;
+    }
+    Local<Value> callback_arg = lock_info_obj;
+    Local<Value> callback_result;
+    {
+      TryCatchScope try_catch_scope(env);
+      if (!grantable_request->callback()
+               ->Call(context, Undefined(isolate), 1, &callback_arg)
+               .ToLocal(&callback_result)) {
+        grantable_request->waiting_promise()
+            ->Reject(context, try_catch_scope.Exception())
+            .Check();
+        grantable_request->released_promise()
+            ->Reject(context, try_catch_scope.Exception())
+            .Check();
+        return;
+      }
+    }
+
+    // Allocate a LockHolder on the heap to safely manage the lock's lifetime
+    // until the user's callback promise settles.
+    auto* fulfill_holder = new LockHolder(granted_lock);
+    auto* reject_holder = new LockHolder(granted_lock);
+    Local<Function> on_fulfilled_callback;
+    Local<Function> on_rejected_callback;
+
+    if (!Function::New(context,
+                       OnLockCallbackFulfilled,
+                       External::New(isolate, fulfill_holder))
+             .ToLocal(&on_fulfilled_callback) ||
+        !Function::New(context,
+                       OnLockCallbackRejected,
+                       External::New(isolate, reject_holder))
+             .ToLocal(&on_rejected_callback)) {
+      delete fulfill_holder;
+      delete reject_holder;
+      return;
+    }
+
+    // Handle promise chain
+    if (callback_result->IsPromise()) {
+      Local<Promise> promise = callback_result.As<Promise>();
+      if (promise->State() == Promise::kRejected) {
+        Local<Value> rejection_value = promise->Result();
+        grantable_request->waiting_promise()
+            ->Reject(context, rejection_value)
+            .Check();
```

Comment:
Not something to do here, but using `Check()` here has the same issue as using `ToLocalChecked()` in that it will just crash the process rather than propagate the error. This is a common issue throughout the code, however, so not something I would block this PR on. We need to handle these better in general. 

If you did want to handle this here, then changing these to check if the return value is empty then doing some proper error propagation similar to the way the ToLocal(...) results are handled would be ideal.